### PR TITLE
[sled-agent] Don't block InstanceManager on full MPSCs

### PR DESCRIPTION
**Pull request by hawkw**
Sled-agent's `InstanceManager` task is responsible for managing the
table of all instances presently running on the sled. When the
sled-agent receives a request relating to an individual instance on the
sled, it's sent to the `InstanceManager` over a `tokio::sync::mpsc`
channel, and is then dispatched by the `InstanceManager` to the
`InstanceRunner` task responsible for that individual instance by
sending it over a *second* `tokio::sync::mpsc` channel. This is where
things start to get interesting.[^1]

`tokio::sync::mpsc` is a *bounded* channel: there is a maximum number of
messages which may be queued by a given MPSC channel at any given time.
The `mpsc::Sender::send` method is an `async fn`, and if the channel is
at capacity, that method will _wait_ until there is once again space in
the channel to send the message being sent. Presently,
`mpsc::Sender::send` is called by the `InstanceManager`'s main run loop
when dispatching a request to an individual instance. As you may have
already started to piece together, this means that if a given
`InstanceRunner` task is not able to process requests fast enough to
drain its channel, the entire `InstanceManager` loop will wait when
dispatching a request to that instance until the queue has been drained.
This means that if one instance's runner task has gotten stuck on
something, like waiting for a Crucible flush that will never complete
(as seen in #6911), that instance will prevent requests being dispatched
to *any other instance* managed by the sled-agent. This is quite
unfortunate!

This commit fixes this behavior by changing the functions that send
requests to an individual instance's task to instead *shed load* when
that instance's request queue is full. We now use the
`mpsc::Sender::try_send` method, rather than `mpsc::Sender::send`, which
does not wait and instead immediately returns an error when the channel
is full. This allows the `InstanceManager` to instead return an error to
the client indicating the channel is full, and move on to processing
requests to other instances which may not be stuck. Thus, a single stuck
instance can no longer block requests from being dispatched to other,
perfectly fine instances.

The error returned when the channel is at capacity is converted to an
HTTP 503 Service Unavailable error by the API. This indicates to the
client that their request to that instance was not able to be processed
at this time, but that it may be processed successfully in the
future.[^2] Now, we can shed load while allowing clients to retry later,
which seems much better than the present situation.

[^1]: In the sense of "may you live in interesting times", naturally.
[^2]: I also considered returning 429 Too Many Requests here, but my
    understanding is that that status code is supposed to indicate that
    too many requests have been received from *that specific client*. In
    this case, we haven't hit a per-client rate limit; we're just
    overloaded by requests more broadly, so it's not that particular
    client's fault.

### CODE REVIEW

**Comment by hawkw** [comment]
Since none of this code is really  meaningfully exercised by the Omicron test suite (due to the pervasive use of the sled-agent simulator), I'd like to do some manual testing of this change (maybe using `omicron-stress`) before merging it. However, I figured I'd go ahead and put the PR up for review now. 

This should hopefully fix the highest-level symptoms described in #6911, but I'm not sure if I want to have this PR close that issue, since the root cause (Crucible getting stuck when trying to perform a snapshot after a disk has disappeared) still needs to be addressed. Also, a potential follow-up from this that I'd like to address is changing how instances are terminated, so that there's a separate mechanism for terminating an instance that takes priority over whatever else it might be doing, which would have better addressed the specific issue in #6911 where a request to expunge the failed disk could not proceed because the instance's task had gotten wedged waiting for the Crucible snapshot to finish, which was unable to complete...because of the bad disk.

**Comment by hawkw** [comment]
Hi @smklein! :)

File: sled-agent/src/instance.rs
Comment (hawkw) [inline sled-agent/src/instance.rs:1153]: The intent behind deriving `Clone` for this is to allow the instance's handle to be cloned  into tasks spawned by the `InstanceManager` for the functions that need to send multiple requests to the instance (or requests to multiple instances) rather than just passing in the `tx` from the original request (e.g. `use_only_these_disks`). I didn't actually do that in this PR but intend to do so in a follow-up.

File: sled-agent/src/instance_manager.rs
Comment (hawkw) [inline sled-agent/src/instance_manager.rs:694]: While working on this change, I noticed that the error handling in the rest of this function is a little bit off. We presently return errors from `create_zone_bundle` if the provided zone name is invalid, or if there's no instance for that UUID. These errors are bubbled up into the `InstanceManager` run loop, which logs an error...but the response `tx` oneshot just gets dropped, rather than receiving these errors.

This means that the client gets a generic "response dropped" error rather than the more useful error indicating that the zone name is bad or doesn't exist. We should change this function so the errors are sent back over the channel instead, but I felt like that was better done in a separate PR from the rest of this change, for Git hustory reasons

File: sled-agent/src/instance_manager.rs
Comment (smklein) [inline sled-agent/src/instance_manager.rs:814]: Here and in "instance.terminate" below are two spots where we still be seem to block on a per-instance basis.

I don't think that needs to block this change -- it's still a move in the right direction -- but "avoid .await-ing on a single instance" seems like necessary follow-up, otherwise we could still hit the https://github.com/oxidecomputer/omicron/issues/6911 issue, right?

(I see you have the `TODO` here, but just confirming that this also applies to `.terminate`)

**Review by smklein — APPROVED** [review]

File: sled-agent/src/instance.rs
Comment (smklein) [reply sled-agent/src/instance.rs:1153]: as a *tiny* nitpick I might punt on deriving clone until that PR, because it doesn't seem necessary in this one.

(but also, not a big deal if this is a pain in the butt)

File: sled-agent/src/instance_manager.rs
Comment (hawkw) [reply sled-agent/src/instance_manager.rs:814]: Yeah, that's correct. I was planning to do that in another branch, but I'm happy to do all of that here, if you prefer?

Comment (smklein) [reply sled-agent/src/instance_manager.rs:814]: Happy to punt it to another PR, if we're doing it somewhere!

File: sled-agent/src/instance_manager.rs
Comment (gjcolombo) [inline sled-agent/src/instance_manager.rs:810]: This might be a better comment on the follow-up PR, but I'm having trouble envisioning how this will work. Today this routine won't return until it's disposed of every VMM in `jobs`. Delegating to individual tasks will allow this loop to move on to consider VMM Y even if requests to VMM X aren't currently being serviced, but won't this routine still need to join all of the tasks at the end?

Not a blocker for this PR; just trying to dispel the "what am I missing?" feeling :)

**Review by gjcolombo — APPROVED** [review]

File: sled-agent/src/instance_manager.rs
Comment (hawkw) [reply sled-agent/src/instance_manager.rs:810]: I was imagining that the instances that must be disposed of would be cloned into a task that would also own the request and it would just go and do the whole thing and then complete the request when it finishes, so that instances which don't need to be destroyed can go on handling requests to do other stuff whilst the disk expungement is being handled. This is in contrast to forcing requests to instances which know nothing about the expunged disk to wait until the expungement has finished. Does that make sense?

Comment (gjcolombo) [reply sled-agent/src/instance_manager.rs:810]: It does, thanks!

The bit I was headed toward is that this ultimately means it will be possible to have an instance manager servicing two separate "use only these disks" commands at once, which is new behavior. That may be fine (I haven't thought about it) but I wanted to make sure I understood what was happening before actually raising it as something worth considering.

**Comment by hawkw** [comment]
Ran `omicron-stress` against this branch (c6035fdd788766e6762799aa812d9ac1eab02be6) for about a minute and a half and eventually hit this:
```
2024-10-23T21:01:34.795972Z ERROR omicron_stress: 202: invalid actor state
2024-10-23T21:01:34.796049Z  INFO omicron_stress: 218: Halting actors
```

which...I'll have to go look into what that actually means --- I'm not sure if that's indicative of an internal bug in `omicron-stress` or something changed in this branch.

**Comment by hawkw** [comment]
Incidentally, the [CI failure on c6035fdd788766e6762799aa812d9ac1eab02be6](https://github.com/oxidecomputer/omicron/pull/6913/checks?check_run_id=31967443465) sure _looks_ like the same Oximeter re-registration issue as #6895/#6916, although it looks like it happened while running a different test: https://buildomat.eng.oxide.computer/wg/0/details/01JAXAGXWPE822857D5D1DP0N4/DMz6rNUPg3AtDtKClnW7rWsxzHSSFztK1IEKxJe1w21To9FJ/01JAXAH9TM5NJ0PDYF6RSJXV5H#S5672

I'm going to re-run that and figure out if it's sticky.

**Comment by hawkw** [comment]
> Ran `omicron-stress` against this branch ([c6035fd](https://github.com/oxidecomputer/omicron/commit/c6035fdd788766e6762799aa812d9ac1eab02be6)) for about a minute and a half and eventually hit this:
> 
> ```
> 2024-10-23T21:01:34.795972Z ERROR omicron_stress: 202: invalid actor state
> 2024-10-23T21:01:34.796049Z  INFO omicron_stress: 218: Halting actors
> ```
> 
> which...I'll have to go look into what that actually means --- I'm not sure if that's indicative of an internal bug in `omicron-stress` or something changed in this branch.

ah, cool, it looks like "invalid actor state" has a more detailed error message string inside it, which the `Display` impl just...eats: https://github.com/oxidecomputer/omicron-stress/blob/8cf99c4520cc861f89f6bcb15a5944c2c7fcc581/src/actor/mod.rs#L52-L53

I'll fix that and hopefully it'll tell us what's actually going on.

**Comment by hawkw** [comment]
aha:
```
2024-10-23T21:25:17.983062Z ERROR omicron_stress: 202: invalid actor state: instance inst1 unexpectedly in state Failed
2024-10-23T21:25:17.983160Z  INFO omicron_stress: 218: Halting actors
```

and if I grep for everything that happened to `inst1`, the last bit immediately before the "invalid state" looks like this:
```
2024-10-23T21:25:08.354247Z  INFO actor{name="inst1_0"}:antagonize{instance_name="inst1"}: omicron_stress::actor::instance: 167: sending instance stop request
2024-10-23T21:25:08.571988Z  INFO actor{name="inst1_7"}:antagonize{instance_name="inst1"}: omicron_stress::actor::instance: 148: sending instance start request
2024-10-23T21:25:11.124327Z  INFO actor{name="inst1_4"}:antagonize{instance_name="inst1"}: omicron_stress::actor::instance: 128: sending instance create request body=InstanceCreate { description: "inst1", disks: [], external_ips: [], hostname: Hostname("inst1"), memory: ByteCount(1073741824), name: Name("inst1"), ncpus: InstanceCpuCount(1), network_interfaces: None, ssh_public_keys: None, start: true, user_data: "" }
2024-10-23T21:25:12.029219Z  WARN actor{name="inst1_2"}:antagonize{instance_name="inst1"}: omicron_stress::actor::instance: 196: instance delete request returned result=Err(Error Response: status: 400 Bad Request; headers: {"content-type": "application/json", "x-request-id": "fcc1c2cb-d1dd-45c4-947a-9f43ab5a1894", "content-length": "181", "date": "Wed, 23 Oct 2024 21:25:11 GMT"}; value: Error { error_code: Some("InvalidRequest"), message: "cannot delete instance: instance is running or has not yet fully stopped", request_id: "fcc1c2cb-d1dd-45c4-947a-9f43ab5a1894" })
2024-10-23T21:25:12.320203Z  INFO actor{name="inst1_7"}:antagonize{instance_name="inst1"}: omicron_stress::actor::instance: 160: instance start request returned result=Ok(Instance { description: "inst1", hostname: "inst1", id: b9f4d940-0e7f-4f59-8293-4d5bc5360898, memory: ByteCount(1073741824), name: Name("inst1"), ncpus: InstanceCpuCount(1), project_id: 74debae8-1eb6-46b9-b5a1-2620c0c08301, run_state: Starting, time_created: 2024-10-23T21:23:25.173472Z, time_modified: 2024-10-23T21:23:25.173472Z, time_run_state_updated: 2024-10-23T21:25:02.436754Z })
2024-10-23T21:25:13.536644Z  INFO actor{name="inst1_3"}:antagonize{instance_name="inst1"}: omicron_stress::actor::instance: 167: sending instance stop request
2024-10-23T21:25:15.316398Z  INFO actor{name="inst1_6"}:antagonize{instance_name="inst1"}: omicron_stress::actor::instance: 179: instance stop request returned result=Ok(Instance { description: "inst1", hostname: "inst1", id: b9f4d940-0e7f-4f59-8293-4d5bc5360898, memory: ByteCount(1073741824), name: Name("inst1"), ncpus: InstanceCpuCount(1), project_id: 74debae8-1eb6-46b9-b5a1-2620c0c08301, run_state: Starting, time_created: 2024-10-23T21:23:25.173472Z, time_modified: 2024-10-23T21:23:25.173472Z, time_run_state_updated: 2024-10-23T21:25:13.483238Z })
2024-10-23T21:25:15.406186Z  WARN actor{name="inst1_1"}:antagonize{instance_name="inst1"}: omicron_stress::actor::instance: 196: instance delete request returned result=Err(Error Response: status: 400 Bad Request; headers: {"content-type": "application/json", "x-request-id": "6f8c02bf-a32b-45ca-bca0-36ee5f1045ab", "content-length": "181", "date": "Wed, 23 Oct 2024 21:25:14 GMT"}; value: Error { error_code: Some("InvalidRequest"), message: "cannot delete instance: instance is running or has not yet fully stopped", request_id: "6f8c02bf-a32b-45ca-bca0-36ee5f1045ab" })
2024-10-23T21:25:16.219577Z  INFO actor{name="inst1_7"}:antagonize{instance_name="inst1"}: omicron_stress::actor::instance: 128: sending instance create request body=InstanceCreate { description: "inst1", disks: [], external_ips: [], hostname: Hostname("inst1"), memory: ByteCount(1073741824), name: Name("inst1"), ncpus: InstanceCpuCount(1), network_interfaces: None, ssh_public_keys: None, start: true, user_data: "" }
2024-10-23T21:25:17.679867Z  INFO actor{name="inst1_5"}:antagonize{instance_name="inst1"}: omicron_stress::actor::instance: 186: sending instance delete request
2024-10-23T21:25:17.948918Z  INFO actor{name="inst1_0"}:antagonize{instance_name="inst1"}: omicron_stress::actor::instance: 179: instance stop request returned result=Ok(Instance { description: "inst1", hostname: "inst1", id: b9f4d940-0e7f-4f59-8293-4d5bc5360898, memory: ByteCount(1073741824), name: Name("inst1"), ncpus: InstanceCpuCount(1), project_id: 74debae8-1eb6-46b9-b5a1-2620c0c08301, run_state: Failed, time_created: 2024-10-23T21:23:25.173472Z, time_modified: 2024-10-23T21:23:25.173472Z, time_run_state_updated: 2024-10-23T21:25:17.431056Z })
2024-10-23T21:25:17.983062Z ERROR omicron_stress: 202: invalid actor state: instance inst1 unexpectedly in state Failed
2024-10-23T21:25:17.983503Z  INFO actor{name="inst1_0"}: omicron_stress::actor: 176: sending halt request
2024-10-23T21:25:17.983535Z  INFO actor{name="inst1_1"}: omicron_stress::actor: 176: sending halt request
2024-10-23T21:25:17.983565Z  INFO actor{name="inst1_2"}: omicron_stress::actor: 176: sending halt request
2024-10-23T21:25:17.983596Z  INFO actor{name="inst1_3"}: omicron_stress::actor: 176: sending halt request
2024-10-23T21:25:17.983651Z  INFO actor{name="inst1_4"}: omicron_stress::actor: 176: sending halt request
2024-10-23T21:25:17.983769Z  INFO actor{name="inst1_5"}: omicron_stress::actor: 176: sending halt request
2024-10-23T21:25:17.983885Z  INFO actor{name="inst1_6"}: omicron_stress::actor: 176: sending halt request
2024-10-23T21:25:17.983961Z  INFO actor{name="inst1_7"}: omicron_stress::actor: 176: sending halt request
````

I need to think through what this actually means...

**Comment by hawkw** [comment]
> aha:
> 
> ```
> 2024-10-23T21:25:17.983062Z ERROR omicron_stress: 202: invalid actor state: instance inst1 unexpectedly in state Failed
> 2024-10-23T21:25:17.983160Z  INFO omicron_stress: 218: Halting actors
> ```
>
> ...

Okay, update on this, looks like instances are also going to `Failed` on the latest `main` commit from yesterday (0f8db77b45a231456d00b77bbc6998e5dfd4f684), so that's not related to this branch. I'm guessing that's related to the `crucible-upstairs` panic that @augustuswm and I observed on the colo rack yesterday, but I'll try and see if there are any cores sitting around after the stress run to confirm that. In any case, it doesn't seem related to this change, so that's good.

**Comment by hawkw** [comment]
> Incidentally, the [CI failure on c6035fdd788766e6762799aa812d9ac1eab02be6](https://github.com/oxidecomputer/omicron/pull/6913/checks?check_run_id=31967443465) sure _looks_ like the same Oximeter re-registration issue as #6895/#6916, although it looks like it happened while running a different test: https://buildomat.eng.oxide.computer/wg/0/details/01JAXAGXWPE822857D5D1DP0N4/DMz6rNUPg3AtDtKClnW7rWsxzHSSFztK1IEKxJe1w21To9FJ/01JAXAH9TM5NJ0PDYF6RSJXV5H#S5672
> 
> I'm going to re-run that and figure out if it's sticky.

Hahahahha, when I [reran that job](https://github.com/oxidecomputer/omicron/pull/6913/checks?check_run_id=31974686692), it failed with a similar Oximeter timeout in _another_ Oximeter test (`integration_tests::oximeter::test_oximeter_database_records`): https://buildomat.eng.oxide.computer/wg/0/details/01JAXAGXWPE822857D5D1DP0N4/DMz6rNUPg3AtDtKClnW7rWsxzHSSFztK1IEKxJe1w21To9FJ/01JAXKSQHRAC3MSG0YH2DBMECP#S5660

I'm still pretty sure that's not my fault...

File: sled-agent/src/instance.rs
Comment (hawkw) [reply sled-agent/src/instance.rs:1192]: I think we can actually just remove this, rather than `Arc`ing it: we never actually use it at present. But, if we want to keep it around so that these can be aborted in future, maybe it's worth it.

```suggestion
    /// This is reference-counted so that the `Instance` struct may be cloned.
```

**Comment by hawkw** [comment]
Argh, an entirely new CI failure appears: https://github.com/oxidecomputer/omicron/pull/6913/checks?check_run_id=32025709910

I really hope this one isn't my fault, but will have to dig a bit to find out:
https://buildomat.eng.oxide.computer/wg/0/details/01JAZVE3ECZRGC5PTJZJB1KJ2N/hh7z1TRG37lRfhcfB0Fy098oj4khSbUKkzckUQR55kd8d4jk/01JAZVEN0ZV92CQJMAQ6X6RX6N#S6429

**Comment by hawkw** [comment]
AHAHAHAHHAHAHA cool I just saw a _new_ Oximeter related test flake: https://buildomat.eng.oxide.computer/wg/0/details/01JAZVE3ECZRGC5PTJZJB1KJ2N/hh7z1TRG37lRfhcfB0Fy098oj4khSbUKkzckUQR55kd8d4jk/01JAZVEN0WFKG1KABT2YD0M74P#S6000

**Pull request merged**
